service_grpc_influxrpc/src/response_chunking.rs

```
@@ -0,0 +1,302 @@
+use observability_deps::tracing::warn;
+
+/// Chunk given [`ReadResponse`]s -- while preserving the [`Frame`] order -- into responses that shall at max have the
+/// given size.
+pub struct ChunkReadResponses<E>
+where
+    E: Send + 'static,
+{
+    inner: BoxStream<'static, Result<Frame, E>>,
+    size_limit: usize,
+    finished: bool,
+    frames: Vec<Frame>,
+    frames_size: usize,
+}
+
+impl<E> ChunkReadResponses<E>
+where
+    E: Send + 'static,
+{
+    /// Create new stream wrapper.
+    ///
+    /// # Panic
+    /// Panics if `size_limit` is 0.
+    pub fn new<S>(inner: S, size_limit: usize) -> Self
+    where
+        S: Stream<Item = Result<ReadResponse, E>> + Send + 'static,
+    {
+        assert!(size_limit > 0, "zero size limit");
+
+        Self {
+            inner: inner
+                .flat_map(|res| match res {
+                    Ok(read_response) => {
+                        futures::stream::iter(read_response.frames).map(Ok).boxed()
+                            as BoxStream<'static, Result<Frame, E>>
+                    }
+                    Err(e) => futures::stream::once(async move { Err(e) }).boxed()
+                        as BoxStream<'static, Result<Frame, E>>,
+                })
+                .boxed(),
+            size_limit,
+            finished: false,
+            frames: Vec::default(),
+            frames_size: 0,
+        }
+    }
+}
+
+impl<E> Stream for ChunkReadResponses<E>
+where
+    E: Send + 'static,
+{
+    type Item = Result<ReadResponse, E>;
+
+    fn poll_next(mut self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Option<Self::Item>> {
+        let this = &mut *self;
+        if this.finished {
+            return Poll::Ready(None);
+        }
+
+        loop {
+            match ready!(this.inner.poll_next_unpin(cx)) {
+                Some(Ok(frame)) => {
+                    let fsize = frame_size(&frame);
+
+                    if fsize > this.size_limit {
+                        warn!(
```

Would it be better to return an error in this case? Or maybe make it a `debug!` message? I was thinking, if we see this message in the logs, is it clear what we would do about it 🤔

---

Yeah, I think we should error in this case. Let me change that.

---

done